File: src/main/java/fit/nlu/tmdt/modules/review/service/ReviewService.java
package fit.nlu.tmdt.modules.review.service;

import fit.nlu.tmdt.modules.review.dto.request.CreateReviewRequest;
import fit.nlu.tmdt.modules.review.dto.request.UpdateReviewRequest;
import fit.nlu.tmdt.modules.review.dto.response.ReviewResponse;
import fit.nlu.tmdt.modules.review.dto.response.ReviewStatsResponse;

import java.util.List;

/**
 * Review Service Interface
 */
public interface ReviewService {

    /**
     * Create review for a post
     */
    ReviewResponse createReview(CreateReviewRequest request, Long userId);

    /**
     * Update review
     */
    ReviewResponse updateReview(Long reviewId, UpdateReviewRequest request, Long userId);

    /**
     * Delete review
     */
    void deleteReview(Long reviewId, Long userId);

    /**
     * Get reviews by post
     */
    List<ReviewResponse> getPostReviews(Long postId, int page, int size);

    /**
     * Get user's reviews
     */
    List<ReviewResponse> getUserReviews(Long userId);

    /**
     * Get landlord's reviews
     */
    List<ReviewResponse> getLandlordReviews(Long landlordId);

    /**
     * Get reviews by landlord ID
     */
    List<ReviewResponse> getReviewsByLandlordId(Long landlordId);

    /**
     * Get average rating for post
     */
    ReviewStatsResponse getAverageRating(Long postId);

    /**
     * Get review by ID
     */
    ReviewResponse getReviewById(Long reviewId);

    /**
     * Landlord response to review
     */
    ReviewResponse landlordRespond(Long reviewId, String response, Long landlordId);

    /**
     * Get review stats for post
     */
    ReviewStatsResponse getPostReviewStats(Long postId);

    /**
     * Mark review as helpful
     */
    void markAsHelpful(Long reviewId, Long userId);

    /**
     * Report review
     */
    void reportReview(Long reviewId, String reason, Long reporterId);
}

File: frontend/src/services/reviewService.ts
import api from './api';
import type { ApiResponse, PaginatedData, ReviewResponse } from '@/types';

export const reviewService = {
  /**
   * Get reviews for a specific post
   */
  getReviewsByPost: async (postId: number | string, page: number = 0, size: number = 10): Promise<PaginatedData<ReviewResponse>> => {
    const response = await api.get<ApiResponse<PaginatedData<ReviewResponse>>>(`/v1/reviews/post/${postId}`, {
      params: { page, size },
    });
    return response.data.data!;
  },

  /**
   * Create a new review
   */
  createReview: async (data: {
    postId: number;
    bookingId?: number;
    rating: number;
    comment: string;
    images?: string[];
  }): Promise<ReviewResponse> => {
    const response = await api.post<ApiResponse<ReviewResponse>>('/v1/reviews', data);
    return response.data.data!;
  },

  /**
   * Delete a review
   */
  deleteReview: async (id: number | string): Promise<void> => {
    await api.delete(`/v1/reviews/${id}`);
  },

  /**
   * Reply to a review (Landlord only)
   */
  replyToReview: async (id: number | string, response: string): Promise<ReviewResponse> => {
    const apiResponse = await api.put<ApiResponse<ReviewResponse>>(`/v1/reviews/${id}/reply`, { response });
    return apiResponse.data.data!;
  },
};

export default reviewService;
